spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java

```diff
@@ -196,4 +198,40 @@ protected static Object convertConstant(Type type, Object value) {
     }
     return value;
   }
+
+  protected void findMoreCompoundConstants(Schema tableSchema, Map<Integer, ?> idToConstant) {
+    List<Types.NestedField> columns = tableSchema.columns();
+
+    for (Types.NestedField nestedField : columns) {
+      visitCompoundTypesByDFS(nestedField, (Map<Integer, Object>) idToConstant);
+    }
+  }
+
+  protected void visitCompoundTypesByDFS(
+          Types.NestedField nestedField,
+          Map<Integer, Object> idToConstant) {
+    switch (nestedField.type().typeId()) {
+      case STRUCT:
```

Comment:
In Iceberg, we separate schema traversal logic from other logic using visitors. I think that you should use a type visitor to implement this rather than traversing a type tree and doing some task in the same recursive function.

Reply:
This is a good idea for refactoring, thanks.